server/src/main/java/org/apache/druid/client/indexing/ClientCompactionRunnerInfo.java

```diff
@@ -202,7 +229,7 @@ public static CompactionConfigValidationResult validateMetricsSpecForMSQ(Aggrega
                  .findFirst()
                  .map(aggregatorFactory ->
                           CompactionConfigValidationResult.failure(
-                              "MSQ: Non-idempotent aggregator[%s] not supported in 'metricsSpec'.",
+                              "MSQ: Non-idempotent aggregator[%s] not supported in 'metricsSpec'",
```

Comment:
Can we update this message to indicate what we mean by non-idempotent?
Maybe a message like `MSQ: Aggregator[%s] is not supported. Aggregation function must be identical to combining function.`

Reply:
Hmm, I am not sure if we should introduce a new term "idempotent aggregators" in the docs which doesn't have any mention in the code (except for MSQ auto-compact validation, AFAIK). So, having "non-idempotent aggregator" in the message might throw off even developers.

Also, is "idempotent" really the word we are looking for? These are aggregators whose combining factory is the same as the aggregator function itself, i.e. combining two partial aggregates would give the same result as aggregating all the values individually. I am not sure if that qualifies for idempotence.

Combining factory is a term which devs know, so at least that's a plus. But I agree that users wouldn't be familiar with it. So, we could even just simplify the message to say that `Aggregator[%s] of type[%s] is not supported.` where the type could be aggregator factory class name. And in the docs, we could have a list of supported aggregators, I guess.

Reply:
Thanks for the comment -- it got me thinking! With the check for `A == A.combiningFactory()`, we are looking to get 2 requirements satisfied:

1. The _function_ to combine partial aggregates is the _same_ as the function to combine individual values. e.g. sum, min ...
2. Reapplying the same _aggregator_ repeatedly on an existing aggregate doesn't change its value: _idempotence_.

The way all druid combining factories are designed, they implicitly satisfy both.  A case where 1 is satisfied but not 2 would be `longSum(name=sum_visit, field=visit)`: input field `visit` could have individual values or partial sums, both would give the same result, but given that `visit` ceases to exist post this operation, reapplying doesn't retain the previous result, so _idempotence_ is violated. An example where 2 is satisfied but not 1 is `average` (not in druid).

Listing out aggregators isn't enough because, similar to the example I gave for longSum, it can have a supported _function_, it can still fail due to the the user-defined properties of the aggregator. But I agree idempotence is just part of the story, and we should just say unsupported aggregator in the code and update the docs a different title and perhaps some clearer explanation.

Reply:
> we should just say unsupported aggregator in the code and update the docs a different title and perhaps some clearer explanation.

Yeah, makes sense to me.

> 2. Reapplying the same _aggregator_ repeatedly on an existing aggregate doesn't change its value: _idempotence_.

In what cases can this be violated? Can we try to capture some examples in the docs?

Reply:
Already cover it in that [section](https://github.com/apache/druid/blob/790dcf0b9a5b5b8ebaf3b109c282218af9c06377/docs/data-management/automatic-compaction.md#idempotent-aggregators). The only case is when input and output columns are different.

Reply:
`count` aggregator btw violates both. count([a,b]) = 2 but count(count([a,b])) = 1. And count outputs value to a separate `count` column instead of to the input columns it is counting.